### PR TITLE
Remove ONLY clause from worker queries

### LINKED ISSUE #475: 'ONLY' clause prevents using inherited tables on worker nodes
Distributed queries to worker nodes include the ONLY clause. This prevents using table inheritance on worker node tables.  Queries in the 5.0 release did not have this clause and worked as expected with inherited tables.  I suspect this is a side effect of pushdown support using the default postgres planner and the master node has no way of knowing that the worker table is inherited.  

The logic for inserting ONLY comes from the ruleutils_9*.c files afaik. As a quick hack to make this work I was able to edit this line:
https://github.com/citusdata/citus/blob/master/src/backend/distributed/utils/ruleutils_95.c#L405

So that "ONLY" is never used.  I assume ONLY has a performance benefit for queries on non inherited tables, but I don't know how impactful it is.

Seems like there are two options here to allow inherited worker tables to work.
1. Never include the ONLY clause in distributed queries
2. Add a flag to the metadata tables to indicate if the worker table has inheritance.

### PR DESCRIPTION
**Pull request by byucesoy**
Fixes #475 

With this change we prevent addition of ONLY clause to queries prepared for
worker nodes. When we add ONLY clause we may miss the inherited tables in
worker nodes created by users manually.

### CODE REVIEW

**Comment by marcocitus** [comment]
This doesn't seem like the right fix. Instead, can you ensure rte->inh is false?

**Comment by byucesoy** [comment]
Did you mean that ensure rte->inh id is true? Because if I set it to false we will add ONLY in every case.

I can do that but in that way we tell particular table is inherited while it may be not. It seems more logical to just removing the check and never add ONLY.

Do you think setting rte->inh to true makes more sense?

**Comment by metdos** [comment]
Hey @marcocitus, would you like to be official reviewer of this PR because you commented on this and it's very small. If you don't have time, that's okay, we will review it here in Istanbul office.

**Comment by mtuncer** [comment]
I would also go for @marcocitus 's suggestion. We barrow **ruleutils** code from postgresql.  looks like setting ```rte->inh``` to true is a better way to go.  We would not need to apply this once we get ruleutils_96.c.

Also, this change only fixes the problem for postgresql 9.5, issue stays still for postgresql 9.4.

**Comment by marcocitus** [comment]
> Did you mean that ensure rte->inh is true? 

Sorry, yes, true.

> I can do that but in that way we tell particular table is inherited while it may be not. 

This usage seems consistent with the PostgreSQL definition:  "inh is TRUE for relation references that should be expanded to include inheritance children, if the rel has any."

> would you like to be official reviewer of this PR 

Sure.

**Comment by byucesoy** [comment]
@marcocitus @mtuncer I modified this change based on your suggestions, now we set inh flag to true before calling `RouterSelectTask` (for SELECT queries) or `RouterModifyTask` (for UPDATE & DELETE queries) so this change removes ONLY clause from SELECT, UPDATE and DELETE queries.

**Comment by marcocitus** [comment]
Thanks. I think a slightly better place for this code would be in RouterSelectTask and RouterModifyTask, since that's where we manipulate the query tree before deparsing it. Could you define a new function which does this transformation and call it from both places?

File: src/backend/distributed/planner/multi_router_planner.c [hunk 1]
```diff
@@ -88,7 +88,24 @@ MultiRouterPlanCreate(Query *query)
...
 	CmdType commandType = query->commandType;
+	ListCell *listCell = NULL;
 	bool modifyTask = false;
+	bool rangeTableInherited = true;
+
+	/*
+	 * We extract RangeTblEntry entries from joinTreeNode and set their inh flag
+	 * to true so that deparser will not add ONLY clause to them.
+	 */
+	foreach(listCell, query->jointree->fromlist)
```
Comment (marcocitus) [inline src/backend/distributed/planner/multi_router_planner.c]: query->rtable gives you a list of RangeTblEntry's that you can use directly.

**Comment by byucesoy** [comment]
I created a function named `RouterSetRangeTablesInherited` for setting inh flag to true of all range table entries. It is called in `RouterSelectTask` and `RouterModifyTask`. Do you guys have better suggestion for name?

Also in this function I set inh flag of all range table entries but would it be better to set inh flag to true only if `RTEKind` is `RTE_RELATION`? Because PostgreSQL does not add ONLY keyword if `RTEKind` is not `RTE_RELATION`?

**Comment by marcocitus** [comment]
> Do you guys have better suggestion for name?

I think for this type of function, which does something to a parameter, we usually start with the verb, e.g. SetRangeTablesInherited. 

> Also in this function I set inh flag of all range table entries but would it be better to set inh flag to true only if RTEKind is RTE_RELATION?

Yes, inh has no meaning for other kinds. 

Note: instead of git merge we usually use "git rebase origin/master" to avoid having a ton of "Merge branch ..." commits, plus it works a bit better.

File: src/backend/distributed/planner/multi_router_planner.c (hunk 1 above)
Comment (marcocitus) [inline src/backend/distributed/planner/multi_router_planner.c]: This code can be removed now.

**Comment by byucesoy** [comment]
I changed the function name and modified the code to set inh flag to true only if rtekind of the range table entry is RTE_RELATION.

File: src/backend/distributed/planner/multi_router_planner.c [hunk 2]
```diff
@@ -1043,3 +1055,24 @@ ColumnMatchExpressionAtTopLevelConjunction(Node *node, Var *column)
 
 	return false;
 }
+
```
Comment (marcocitus) [inline src/backend/distributed/planner/multi_router_planner.c]: We typically leave 2 empty lines between functions. You can use the citus_indent tool to automatically enforce our code style.

File: src/backend/distributed/planner/multi_router_planner.c [hunk 3]
```diff
@@ -1043,3 +1055,24 @@ ColumnMatchExpressionAtTopLevelConjunction(Node *node, Var *column)
...
+	List *rangeTableList = query->rtable;
+	ListCell *rangeTableCell = NULL;
+	bool rangeTableEntryInherited = true;
+
+	foreach(rangeTableCell, rangeTableList)
+	{
+		RangeTblEntry *rangeTableEntry = (RangeTblEntry *) lfirst(rangeTableCell);
+		if (rangeTableEntry->rtekind == RTE_RELATION)
+		{
+			rangeTableEntry->inh = rangeTableEntryInherited;
```
Comment (marcocitus) [inline src/backend/distributed/planner/multi_router_planner.c]: I think using true directly would be slightly more clear here.

File: src/backend/distributed/planner/multi_router_planner.c [hunk 4]
```diff
@@ -71,7 +71,7 @@ static Const * ExtractInsertPartitionValue(Query *query, Var *partitionColumn);
 static Task * RouterSelectTask(Query *query);
 static Job * RouterQueryJob(Query *query, Task *task);
 static bool ColumnMatchExpressionAtTopLevelConjunction(Node *node, Var *column);
-
+static void SetRangeTablesInherited(Query *query);
```
Comment (marcocitus) [inline src/backend/distributed/planner/multi_router_planner.c]: Should have 2 empty lines here.

**Comment by marcocitus** [comment]
After minor style fixes and reebase (git rebase origin/master), this looks good to go.

**Comment by jasonmp85** [comment]
Hey all… this was marked as a `fix`. Do we need it to go into a release (5.1.1) or something, or is it just a minor fix for `master` only?

**Pull request merged**